pkg/profilecreator/profilecreator.go

```diff
@@ -312,6 +307,34 @@ func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, sp
 
 }
 
+// getCPUsSequentially returns Reserved and Isolated CPUs sequentially
+func (ghwHandler GHWHandler) getCPUsSequentially(reservedCPUCount int, topologyInfoNodes []*topology.Node) (string, string, error) {
+	totalCPUSet := cpuset.NewBuilder()
+	reservedCPUSet := cpuset.NewBuilder()
+	max := reservedCPUCount
+	htEnabled, err := ghwHandler.isHyperthreadingEnabled()
+	if err != nil {
+		return "", "", fmt.Errorf("Error determining if Hyperthreading is enabled or not: %v", err)
+	}
+	if max%2 != 0 && htEnabled {
+		return "", "", fmt.Errorf("Can't allocatable odd number of CPUs from a NUMA Node")
+	}
+	for _, node := range topologyInfoNodes {
+		for _, processorCores := range node.Cores {
+			for _, core := range processorCores.LogicalProcessors {
+				totalCPUSet.Add(core)
+				if reservedCPUSet.Result().Size() < max {
```

Comment:
this can be a head-scratcher. Per `cpuset.Builder` docs, `Result()` should be the last operation performed in a builder, meaning that after a `Result()` call, all following `Add()` calls should be silently ignored.

However please note that the method receivers of the `Builder` type are all by value (https://tour.golang.org/methods/8). This means that:
1. when `Result` is called, a *copy* of `reservedCPUSet` is passed as method receiver of `Result`
2. `Result` marks the *copy* it received (not the original object, `reservedCPUSet` in our case) as done (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpuset/cpuset.go#L60)
3. We call `Size` on the modified copy. This yields the expected results
4. the modified *copy* is discarded
5. the original object (reservedCPUSet) is unchanged, we can safely Add() cores to it.

This probably means the API of cpuset.Builder could be improved, but I also wonder if we should rearchitect our function to better reflect the intended usage of `cpuset.Builder`, on which `Result()` should be the very last operation performed on them.

Reply:
Thanks for the great insight and the detailed explanation here. Given that cpuset.Builder is a mutable (and cpuset.CPUSet is immutable) we had to take the route of using cpuset.Builder. Also we have to check the size of the CPUset inside the Builder (for which we have to access Result()) to be able to allocate the correct amount of CPUs from a NUMA node before calling the Add(). Given that the current implementation works fine, I am going to leave it as is for now but will try to figure out how this can be implemented in a better way but I guess that will be done as a separate KEP when we have a better idea.